src/nl/hannahsten/texifyidea/util/LatexPsiFactory.kt

```diff
@@ -0,0 +1,74 @@
+package nl.hannahsten.texifyidea.util
+
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiFileFactory
+import nl.hannahsten.texifyidea.LatexLanguage
+import nl.hannahsten.texifyidea.psi.LatexCommands
+import nl.hannahsten.texifyidea.psi.LatexEnvironment
+import nl.hannahsten.texifyidea.psi.LatexEnvironmentContent
+
+class LatexPsiFactory(private val project: Project) {
+
+    fun createEnvironmentContent(): LatexEnvironmentContent {
+        val environment = createFromText("\\begin{figure}\n" +
+                "        Placehodler\n" +
+                "    \\end{figure}").firstChildOfType(LatexEnvironment::class)!!
+        environment.environmentContent!!.firstChild.delete()
+        return environment.environmentContent!!
+    }
+
+    fun createUniqueLabelCommandFor(command: LatexCommands): PsiElement? {
+        val required = command.requiredParameters
+        if (required.isEmpty()) {
+            return null
+        }
+
+        // Determine label name.
+        val prefix = Magic.Command.labeled[command.name]
+        val labelName = required[0].formatAsLabel()
+        val labelBase = "$prefix:$labelName"
+
+        return createUniqueLabelCommand(labelBase, command.containingFile)
+    }
+
+    fun createUniqueLabelCommandFor(command: LatexEnvironment): PsiElement? {
+        // Determine label name.
+        val prefix = Magic.Environment.labeled[command.environmentName]
+        val labelName = command.environmentName.formatAsLabel()
+        val labelBase = "$prefix:$labelName"
+
+        return createUniqueLabelCommand(labelBase, command.containingFile)
+    }
+
+    private fun createUniqueLabelCommand(labelBase: String, file: PsiFile): PsiElement {
+        val allLabels = file.findLabelsInFileSet()
+        val createdLabel = appendCounter(labelBase, allLabels)
+        return createLabelCommand(createdLabel)
+    }
+
+    fun createLabelCommand(labelName: String): PsiElement {
+        val labelText = "\\label{$labelName}"
+        val fileFromText = createFromText(labelText)
+        return fileFromText.firstChild
+    }
+
+    private fun createFromText(text: String): PsiElement =
+            PsiFileFactory.getInstance(project).createFileFromText("DUMMY.tex", LatexLanguage.INSTANCE, text, false, true)
```

Comment:
@slideclimb Didn't you also need to update the PSI tree by creating a new node, for the refactoring support? This class is doing that, if I'm not mistaken.

Reply:
Maybe we can find a way to create the PSI nodes directly instead of parsing short text snippets, but I couldn't find a way to do it.

Reply:
I doubt that. From https://www.jetbrains.org/intellij/sdk/docs/reference_guide/custom_language_support/rename_refactoring.html

> Creating a fully correct AST node from scratch is quite difficult. Thus, surprisingly, the easiest way to get the replacement node is to create a dummy file in the custom language so that it would contain the necessary node in its parse tree, build the parse tree and extract the necessary node from it.